pkg/cmd/openshift-sdn/proxy.go

```diff
@@ -346,3 +352,71 @@ func getNodeIP(client kv1core.CoreV1Interface, hostname string) (net.IP, error)
 
 	return nodeIP, nil
 }
+
+type waitingProxyHandler struct {
+	initialized uint32
+	// WaitChan will be closed when both services and endpoints have
+	// been synced in the proxy
+	waitChan chan<- bool
+
+	serviceChild    pconfig.ServiceHandler
+	serviceSynced   uint32
+	endpointsChild  pconfig.EndpointsHandler
+	endpointsSynced uint32
+}
+
+func newWaitingProxyHandler(serviceChild pconfig.ServiceHandler, endpointsChild pconfig.EndpointsHandler, waitChan chan<- bool) *waitingProxyHandler {
+	return &waitingProxyHandler{
+		serviceChild:   serviceChild,
+		endpointsChild: endpointsChild,
+		waitChan:       waitChan,
+	}
+}
+
+func (wph *waitingProxyHandler) checkInitialized() {
+	if atomic.LoadUint32(&wph.initialized) > 0 {
+		// Don't close our channel twice
+		return
+	}
+	if atomic.LoadUint32(&wph.serviceSynced) > 0 && atomic.LoadUint32(&wph.endpointsSynced) > 0 {
+		klog.V(2).Info("openshift-sdn proxy services and endpoints initialized")
+		atomic.StoreUint32(&wph.initialized, 1)
+		close(wph.waitChan)
+	}
+}
+
+func (wph *waitingProxyHandler) OnServiceAdd(service *v1.Service) {
+	wph.serviceChild.OnServiceAdd(service)
+}
+
+func (wph *waitingProxyHandler) OnServiceUpdate(oldService, service *v1.Service) {
+	wph.serviceChild.OnServiceUpdate(oldService, service)
```

Comment:
During update are we still synced and initialized?

Reply:
Yeah, see above.